File: android/app/src/main/java/fr/jarvis/companion/JarvisApi.java
package fr.jarvis.companion;

import android.content.Context;
import android.os.Build;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/** Client HTTPS natif. Le certificat JARVIS est fourni par network_security_config. */
final class JarvisApi {
    interface Callback { void complete(Result result); }

    static final class Result {
        final boolean ok;
        final int status;
        final JSONObject json;
        final String cookie;
        final String error;

        Result(boolean ok, int status, JSONObject json, String cookie, String error) {
            this.ok = ok;
            this.status = status;
            this.json = json;
            this.cookie = cookie;
            this.error = error;
        }
    }

    private static final ExecutorService NETWORK = Executors.newFixedThreadPool(2);
    private final Context context;

    JarvisApi(Context context) {
        this.context = context.getApplicationContext();
    }

    void completePairing(String code, Callback callback) {
        JSONObject body = new JSONObject();
        try {
            body.put("code", code);
            body.put("device_id", JarvisSettings.deviceId(context));
            body.put("name", "JARVIS sur " + Build.MODEL);
            body.put("model", Build.MANUFACTURER + " " + Build.MODEL);
            body.put("app_version", BuildConfig.VERSION_NAME);
        } catch (Exception ignored) {}
        post("/api/mobile/pairing/complete", body, "", callback);
    }

    void createWebSession(Callback callback) {
        post("/api/mobile/session", new JSONObject(), JarvisSettings.nativeToken(context), callback);
    }

    void registerPushToken(String fcmToken) {
        JSONObject body = new JSONObject();
        try { body.put("token", fcmToken); } catch (Exception ignored) {}
        post("/api/mobile/push-token", body, JarvisSettings.nativeToken(context), result -> {});
    }

    void updateCapabilities(boolean location, boolean wakeWord) {
        JSONObject body = new JSONObject();
        try {
            body.put("push", BuildConfig.FIREBASE_CONFIGURED);
            body.put("background_location", location);
            body.put("wake_word", wakeWord);
        } catch (Exception ignored) {}
        post("/api/mobile/capabilities", body, JarvisSettings.nativeToken(context), result -> {});
    }

    void postLocation(double latitude, double longitude, double altitude,
                      float accuracy, float speed, long timestamp) {
        JSONObject body = new JSONObject();
        try {
            body.put("latitude", latitude);
            body.put("longitude", longitude);
            body.put("altitude", altitude);
            body.put("accuracy", accuracy);
            body.put("speed", speed);
            body.put("timestamp", timestamp);
            body.put("source", "android_background");
        } catch (Exception ignored) {}
        post("/api/location", body, JarvisSettings.nativeToken(context), result -> {});
    }

    private void post(String path, JSONObject body, String bearer, Callback callback) {
        NETWORK.execute(() -> callback.complete(request(path, body, bearer)));
    }

    private Result request(String path, JSONObject body, String bearer) {
        HttpURLConnection connection = null;
        try {
            URL url = new URL(JarvisSettings.server(context) + path);
            connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("POST");
            connection.setConnectTimeout(12_000);
            connection.setReadTimeout(20_000);
            connection.setRequestProperty("Content-Type", "application/json; charset=utf-8");
            connection.setRequestProperty("Accept", "application/json");
            connection.setRequestProperty("User-Agent", "JARVIS-Android/" + BuildConfig.VERSION_NAME);
            if (bearer != null && !bearer.isEmpty()) {
                connection.setRequestProperty("Authorization", "Bearer " + bearer);
            }
            connection.setDoOutput(true);
            byte[] bytes = body.toString().getBytes(StandardCharsets.UTF_8);
            connection.getOutputStream().write(bytes);

            int status = connection.getResponseCode();
            InputStream stream = status >= 200 && status < 400
                    ? connection.getInputStream() : connection.getErrorStream();
            String text = read(stream);
            JSONObject json = text.isEmpty() ? new JSONObject() : new JSONObject(text);
            String cookie = connection.getHeaderField("Set-Cookie");
            String error = json.optString("detail", json.optString("error", "HTTP " + status));
            return new Result(status >= 200 && status < 300, status, json, cookie, error);
        } catch (Exception e) {
            return new Result(false, 0, new JSONObject(), "", e.getMessage());
        } finally {
            if (connection != null) connection.disconnect();
        }
    }

    private String read(InputStream stream) throws Exception {
        if (stream == null) return "";
        StringBuilder result = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) result.append(line);
        }
        return result.toString();
    }
}

File: android/app/src/main/kotlin/fr/jarvis/companion/network/JarvisApi.kt
package fr.jarvis.companion.network

import android.content.Context
import android.os.Build
import fr.jarvis.companion.BuildConfig
import fr.jarvis.companion.data.JarvisSettings
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import javax.net.ssl.HttpsURLConnection

/**
 * Client HTTPS natif vers le serveur JARVIS.
 * Confiance TLS : CA privée JARVIS pour l'hôte configuré (voir [JarvisTls]), CA système ailleurs.
 */
class JarvisApi(context: Context) {
    private val appContext = context.applicationContext

    fun completePairing(code: String, callback: (JarvisApiResult) -> Unit) {
        val body = JSONObject().apply {
            put("code", code)
            put("device_id", JarvisSettings.deviceId(appContext))
            put("name", "JARVIS sur ${Build.MODEL}")
            put("model", "${Build.MANUFACTURER} ${Build.MODEL}")
            put("app_version", BuildConfig.VERSION_NAME)
        }
        post("/api/mobile/pairing/complete", body, bearer = "", callback)
    }

    fun validateNativeToken(callback: (JarvisApiResult) -> Unit) {
        post("/api/mobile/session", JSONObject(), JarvisSettings.nativeToken(appContext), callback)
    }

    fun pingAuthStatus(callback: (JarvisApiResult) -> Unit) {
        get("/api/auth/status", callback)
    }

    fun registerPushToken(fcmToken: String) {
        val body = JSONObject().put("token", fcmToken)
        post("/api/mobile/push-token", body, JarvisSettings.nativeToken(appContext)) {}
    }

    fun updateCapabilities(location: Boolean, wakeWord: Boolean) {
        val body = JSONObject().apply {
            put("push", BuildConfig.FIREBASE_CONFIGURED)
            put("background_location", location)
            put("wake_word", wakeWord)
        }
        post("/api/mobile/capabilities", body, JarvisSettings.nativeToken(appContext)) {}
    }

    fun postLocation(
        latitude: Double,
        longitude: Double,
        altitude: Double,
        accuracy: Float,
        speed: Float,
        timestamp: Long,
    ) {
        val body = JSONObject().apply {
            put("latitude", latitude)
            put("longitude", longitude)
            put("altitude", altitude)
            put("accuracy", accuracy)
            put("speed", speed)
            put("timestamp", timestamp)
            put("source", "android_background")
        }
        post("/api/location", body, JarvisSettings.nativeToken(appContext)) {}
    }

    private fun get(path: String, callback: (JarvisApiResult) -> Unit) {
        NETWORK.execute {
            callback(request(path, method = "GET", body = null, bearer = ""))
        }
    }

    private fun post(
        path: String,
        body: JSONObject,
        bearer: String,
        callback: (JarvisApiResult) -> Unit,
    ) {
        NETWORK.execute {
            callback(request(path, method = "POST", body = body, bearer = bearer))
        }
    }

    private fun request(
        path: String,
        method: String,
        body: JSONObject?,
        bearer: String,
    ): JarvisApiResult {
        var connection: HttpURLConnection? = null
        return try {
            val url = URL(JarvisSettings.server(appContext) + path)
            connection = (url.openConnection() as HttpURLConnection).apply {
                if (this is HttpsURLConnection &&
                    ServerUrlNormalizer.isJarvisHost(
                        JarvisSettings.server(appContext),
                        url.host,
                    )
                ) {
                    val ctx = JarvisTls.sslContext(appContext)
                    sslSocketFactory = ctx.socketFactory
                }
                requestMethod = method
                connectTimeout = CONNECT_TIMEOUT_MS
                readTimeout = READ_TIMEOUT_MS
                setRequestProperty("Accept", "application/json")
                setRequestProperty("User-Agent", "JARVIS-Android/${BuildConfig.VERSION_NAME}")
                if (bearer.isNotEmpty()) {
                    setRequestProperty("Authorization", "Bearer $bearer")
                }
                if (method == "POST" && body != null) {
                    setRequestProperty("Content-Type", "application/json; charset=utf-8")
                    doOutput = true
                    outputStream.use { stream ->
                        stream.write(body.toString().toByteArray(StandardCharsets.UTF_8))
                    }
                }
            }

            val status = connection.responseCode
            val stream = if (status in 200..399) connection.inputStream else connection.errorStream
            val text = readStream(stream)
            val json = if (text.isEmpty()) JSONObject() else JSONObject(text)
            val cookie = connection.getHeaderField("Set-Cookie")
            val error = json.optString("detail", json.optString("error", "HTTP $status"))
            JarvisApiResult(
                ok = status in 200..299,
                status = status,
                json = json,
                cookie = cookie,
                error = error,
            )
        } catch (e: Exception) {
            JarvisApiResult.failure(e.message ?: "erreur réseau")
        } finally {
            if (connection is HttpsURLConnection) {
                connection.disconnect()
            } else {
                connection?.disconnect()
            }
        }
    }

    private fun readStream(stream: InputStream?): String {
        if (stream == null) return ""
        return stream.bufferedReader(StandardCharsets.UTF_8).use(BufferedReader::readText)
    }

    companion object {
        private val NETWORK: ExecutorService = Executors.newFixedThreadPool(2)
        private const val CONNECT_TIMEOUT_MS = 12_000
        private const val READ_TIMEOUT_MS = 20_000
    }
}
